Encryption.md

```diff
@@ -19,216 +19,544 @@
+
+Unique nonces are generated for each ciphertext. The nonce length is 12 bytes (96 bits). The NIST 
+specification requires a “random field” to fill all 12 bytes and an arbitrary “free field” 
+to be empty for this nonce length (section 8.2.2). Parquet uses the option 1 for RGB implementation - 
+"The random field shall .. consist of an output string of r(i) bits from an approved RBG with a 
+sufficient security strength". That is, a fresh random byte array must be generated for each nonce - 
+as opposed to the option 2 in the same spec - "or the result of applying the r(i)–bit incrementing 
+function to the random field of the preceding <nonce> for the given key"
+
+
+### 4.2 AES_GCM_CTR_V1 encryption algorithm
+In this Parquet algorithm, all Thrift modules are encrypted with the GCM cipher, as described above. 
+
+The pages are encrypted by the CTR cipher without padding. The AES CTR cipher
+must be implemented by a cryptographic provider according to the NIST SP 800-38A specification. 
+
+Unique nonces are generated for each ciphertext, with the nonce length of 12 bytes. All 12 
+bytes are constructed with an RBG option 1 as described for the AES_GCM_V1 algorithm. The data 
+module encryption and decryption is performed with a 16 byte IV comprised of the nonce and a 4-byte 
+initial counter field. The first 31 bits of the initial counter field are set to 0, the last bit 
+is set to 1. 
+
+### 4.3 Key metadata
+A wide variety of services and tools for management of encryption keys exist in the 
+industry today. Public clouds offer different key management services (KMS), and 
+organizational IT systems either build proprietary key managers in-house or adopt open source 
+tools for on-premises deployment. Besides the diversity of management tools, there are many 
+ways to generate and handle the keys themselves (generate Data keys inside KMS – or locally 
+upon data encryption; use Data keys only, or use Master keys to encrypt/wrap the Data keys; 
+store the wrapped key material inside the data file, or at a separate location; etc). There 
+is also a large variety of authorization and certification methods, required to control the 
+access to encryption keys.
+
+Parquet is not limited to a single KMS, key generation/wrapping method, or authorization service. 
+Instead, Parquet provides a developer with a simple interface that can be utilized for implementation 
+of any key management scheme. For each column or footer key, a file writing code can pass an 
+arbitrary `key_metadata` byte array that will be stored in the file. A file reading code can pass a 
+key retrieval callback, that will invoked on the key_metadata byte array in order to recover the
+encryption (or signature) key. For example, the key_metadata can keep a serialized
+
+   * String ID of a Data encryption key (for direct key retrieval from a KMS).
+   * Wrapped Data key, and string ID of a Master key. The Data key is generated randomly and 
```

Comment:
By "wrapped" do you mean encrypted?

Reply:
Yep. Will add this in the text.

Reply:
Instead of adding synonyms for "encrypted", just use "encrypted". The more words you use to mean the same thing, the more confusing it is for implementers.